test/functional/feature_checktemplateverify.py

```diff
@@ -0,0 +1,468 @@
+#!/usr/bin/env python3
+# Copyright (c) 2015-2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test (CheckTemplateVerify)
+"""
+
+from test_framework.blocktools import create_coinbase, create_block, add_witness_commitment
+from test_framework.messages import CTransaction, CTxOut, CTxIn, CTxInWitness, COutPoint, COIN, sha256
+from test_framework.p2p import P2PInterface
+from test_framework.script import CScript, OP_TRUE, OP_DEPTH, OP_ENDIF, OP_IF, OP_CHECKTEMPLATEVERIFY, OP_FALSE, OP_DROP, taproot_construct
+from test_framework.script_util import script_to_p2sh_script
+from test_framework.key import ECKey, compute_xonly_pubkey
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal, assert_raises_rpc_error
+from test_framework.wallet import MiniWallet, MiniWalletMode
+from decimal import Decimal
+import random
+from io import BytesIO
+from test_framework.address import script_to_p2sh
+
+CHECKTEMPLATEVERIFY_ERROR = "non-mandatory-script-verify-flag (Script failed an OP_CHECKTEMPLATEVERIFY operation)"
+DISCOURAGED_ERROR = "non-mandatory-script-verify-flag (NOPx reserved for soft-fork upgrades)"
+STACK_TOO_SHORT_ERROR = "non-mandatory-script-verify-flag (Operation not valid with the current stack size)"
+def random_bytes(n):
+    return bytes(random.getrandbits(8) for i in range(n))
+def template_hash_for_outputs(outputs, nIn = 0, nVin=1, vin_override=None):
+    c = CTransaction()
+    c.nVersion = 2
+    c.vin = vin_override
+    if vin_override is None:
+        c.vin = [CTxIn()]*nVin
+    c.vout = outputs
+    return c.get_standard_template_hash(nIn)
+def random_p2sh():
+    return script_to_p2sh_script(random_bytes(20))
+def random_real_outputs_and_script(n, nIn=0, nVin=1, vin_override=None):
+    outputs = [CTxOut((x+1)*1000, random_p2sh()) for x in range(n)]
+    script  = CScript([template_hash_for_outputs(outputs, nIn, nVin, vin_override), OP_CHECKTEMPLATEVERIFY])
+    return outputs, script
+def random_secure_tree(depth):
+    leaf_nodes = [CTxOut(100, CScript(bytes([0, 0x14]) + random_bytes(20))) for x in range(2**depth)]
```

Comment:
Using a for loop instead of list comprehension here *may* slow down the test suite slightly, but might be worth it for enhanced comprehension of anyone reading/maintaining these tests later?

```suggestion
    leaf_nodes = []
    for x in range(2**depth):
        nValue = 100
        scriptPubKey = CScript(bytes([0, 0x14]) + random_bytes(20))
        node = CTxOut(nValue, scriptPubKey)
        leaf_nodes.append(node)
```

Also, my understanding is `bytes([0, 0x14]) + random_bytes(20)` means OP_FALSE and push 20 random bytes onto the stack?

The [wiki](https://en.bitcoin.it/wiki/Script) specifies OP_FALSE as:
> "An empty array of bytes is pushed onto the stack. (This is not a no-op: an item is added to the stack.)`

The 20 bytes seem something other than empty, but my understanding is very basic; I've just now looked at the Wiki to understand what this line of code means. Clarification welcome :).

Reply:
if i recall I just wanted something unique and non-spendable... but it's probably worth documenting better.